database/mongodb/mongodb.go

```diff
@@ -184,10 +207,70 @@ func (m *Mongo) Drop() error {
 	return m.db.Drop(context.TODO())
 }
 
+func (m *Mongo) ensureLockTable() error {
+	indexes := m.db.Collection(m.config.LockingCollection).Indexes()
+	indexOptions := options.Index().SetUnique(true).SetName("lock_unique_key")
+	indexKeys := bsonx.MDoc{
+		DefaultLockingKey: bsonx.Int32(-1),
+	}
+	_, err := indexes.CreateOne(context.TODO(), mongo.IndexModel{
+		Options: indexOptions,
+		Keys:    indexKeys,
+	})
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+	return nil
+}
+
+// ensureVersionTable checks if versions table exists and, if not, creates it.
+// Note that this function locks the database, which deviates from the usual
+// convention of "caller locks" in the MongoDb type.
+func (m *Mongo) ensureVersionTable() (err error) {
+	if err = m.Lock(); err != nil {
+		return err
+	}
+
+	defer func() {
+		if e := m.Unlock(); e != nil {
+			if err == nil {
+				err = e
+			} else {
+				err = multierror.Append(err, e)
+			}
+		}
+	}()
+
+	if err != nil {
+		return err
+	}
+	if _, _, err = m.Version(); err != nil {
+		return err
+	}
+	return nil
+}
+
 func (m *Mongo) Lock() error {
```

Comment:
Should we document anywhere about this approach, and maybe that it is an advisory or cooperative locking. What about that we got it from MongoBee (if you used their approach). I guess the equivalent of javadoc isn't followed in this code base, but there are comments for some algorithmic things.

Reply:
done